### PR TITLE
Correct <input type=checkbox switch> animation corner case and add tests

### PR DESCRIPTION
**Pull request by annevk**
#### c2e7027de483cb5d51dd8a79965e10a4851ebc52
<pre>
Correct &lt;input type=checkbox switch&gt; animation corner case and add tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=265356">https://bugs.webkit.org/show_bug.cgi?id=265356</a>

Reviewed by Aditya Keerthi.

I&apos;ve decided on the following things:

* When an animation is running and setChecked() is called with the same
  value the animation is running for, the animation continues to run.
  As such we do not modify the early return in setChecked() for when
  the state does not change unlike an earlier iteration of this change.
* When an animation is running and the checked or disabled state
  changes underneath, the animation is stopped.

Stopping the animation is abstracted into a method for clarity.

Test coverage is added for these scenarios and a couple more to reduce
the chances of regressions when refactoring in the future.

I also want to mention here that I strongly considered an alternative
approach whereby CheckboxInputType would be solely responsible for
calling performSwitchCheckedChangeAnimation() right after it calls
setChecked(), but setChecked() can invalidate or reverse an animation
and I could not find a good way to integrate that behavior.

Canonical link: <a href="https://commits.webkit.org/271327@main">https://commits.webkit.org/271327@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f193fa4718b7cc7753252facf37f7d9bfcf022d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30430 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25464 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3932 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25233 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28165 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5301 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4564 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4755 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24975 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31118 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25508 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25414 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31009 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4760 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2918 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28820 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6284 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24732 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6722 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5206 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5231 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/html/HTMLInputElement.cpp [hunk 1]
```diff
@@ -1033,7 +1033,7 @@ void HTMLInputElement::setDefaultCheckedState(bool isDefaultChecked)
 void HTMLInputElement::setChecked(bool isChecked, WasSetByJavaScript wasCheckedByJavaScript)
 {
     m_dirtyCheckednessFlag = true;
-    if (checked() == isChecked)
+    if (checked() == isChecked && (!isSwitch() || switchCheckedChangeAnimationProgress() == 1.0f))
         return;
```
Comment (pxlcoder) [inline Source/WebCore/html/HTMLInputElement.cpp]: Are there any implications on event dispatch if we don't early return here during the animation?

Comment (annevk) [reply Source/WebCore/html/HTMLInputElement.cpp]: I don't think so, but I've changed my mind on this particular change. I spend a bunch more time thinking about the overall setup here, but I couldn't really figure out a better alternative. I elaborated on that a bit in the commit message.

Maybe that will come as we add more complexity for iOS and pointer tracking though.

**Review by pxlcoder — APPROVED** [review]
